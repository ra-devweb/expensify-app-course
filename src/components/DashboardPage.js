import React from 'react';

import ExpensesList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilters';

const DashboardPage = () => (
  <div>
    <ExpenseListFilter />
    <ExpensesList />
  </div>
);

export default DashboardPage;
