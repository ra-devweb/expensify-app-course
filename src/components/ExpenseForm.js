import React from 'react';
import { SingleDatePicker } from 'react-dates';
import moment from 'moment';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
    };
  }

  onChangeDescription = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onChangeNote = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onChangeAmount = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onChangeDate = (createdAt) => {
    if (createdAt) this.setState(() => ({ createdAt }));
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmitForm = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount.',
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmitForm}>
          <div>
            <label htmlFor='description'>Description</label>
            <input
              type='text'
              id='description'
              placeholder='Description'
              value={this.state.description}
              onChange={this.onChangeDescription}
              autoFocus
            />
          </div>
          <div>
            <label htmlFor='amount'>Amount</label>
            <input
              id='amount'
              type='text'
              placeholder='Amount'
              value={this.state.amount}
              onChange={this.onChangeAmount}
            />
          </div>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onChangeDate}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <div>
            <label htmlFor='note'>Note (Optional)</label>
            <textarea
              id='note'
              placeholder='Add a note for expense'
              value={this.state.note}
              onChange={this.onChangeNote}
            ></textarea>
          </div>
          <button>Add expense</button>
        </form>
      </div>
    );
  }
}
