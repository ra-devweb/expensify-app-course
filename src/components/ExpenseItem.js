import React from 'react';
import { Link } from 'react-router-dom';

const ExpenseItem = ({ id, description, amount, createdAt }) => {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h1>{description}</h1>
      </Link>
      <p>{amount}</p>
      <span>({createdAt})</span>
    </div>
  );
};

export default ExpenseItem;
