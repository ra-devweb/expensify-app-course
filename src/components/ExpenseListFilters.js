import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate,
} from '../actions/filters';

export class ExpenseListFilter extends React.Component {
  state = {
    focusedInput: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (focusedInput) => {
    this.setState(() => ({
      focusedInput,
    }));
  };

  onTextChange = (e) => this.props.setTextFilter(e.target.value);

  onSortChange = (e) => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className='filters'>
        <div className='container'>
          <div className='form text-center'>
            <div className='form__group'>
              <input
                className='form__input'
                placeholder='Search expenses'
                type='text'
                value={this.props.filters.text}
                onChange={this.onTextChange}
              />
            </div>
            <div className='form__group'>
              <select 
                className='form__select'
                value={this.props.filters.sortBy} onChange={this.onSortChange}
                >
                <option value='date'>Date</option>
                <option value='amount'>Amount</option>
              </select>
            </div>
            <div className='form__group'>
              <DateRangePicker
                startDate={this.props.filters.startDate}
                startDateId='startDateId'
                endDate={this.props.filters.endDate}
                endDateId='endDateId'
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.focusedInput}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters,
});

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
