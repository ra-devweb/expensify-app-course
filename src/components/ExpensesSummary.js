import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import numeral from 'numeral';

import selectExpenses from '../selectors/expenses';
import addExpensesTotal from '../selectors/addExpensesTotal';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';

  const formattedExpensesTotal = numeral(expensesTotal / 100).format(
    '0,0.00 $'
  );

  return (
    <div className='summary'>
      <div className='container'>
        <h1 className='summary__title'>
          Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{formattedExpensesTotal}</span>
        </h1>

        <div className='summary__box-btn'>
          <Link className='summary__link btn' to='/create'>&#9998; Add expense</Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: addExpensesTotal(visibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
