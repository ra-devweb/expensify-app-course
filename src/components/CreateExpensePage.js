import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class CreateExpensePage extends React.Component {
  onSubmit = ({ description, amount, createdAt, note }) => {
    this.props.addExpense({ description, amount, createdAt, note });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addExpense: (expense) => dispatch(addExpense(expense)),
  };
};

export default connect(undefined, mapDispatchToProps)(CreateExpensePage);
