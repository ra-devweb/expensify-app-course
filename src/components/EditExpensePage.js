import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { updateExpense, deleteExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  updateExpense = (expense) => {
    this.props.updateExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  deleteExpense = () => {
    this.props.deleteExpense(this.props.expense.id);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.updateExpense}
        />
        <button onClick={this.deleteExpense}>Remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(
      (expense) => expense.id === props.match.params.id
    ),
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    updateExpense: (id, expense) => dispatch(updateExpense(id, expense)),
    deleteExpense: (id) => dispatch(deleteExpense({ id })),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
