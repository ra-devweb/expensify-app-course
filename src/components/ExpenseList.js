import React from 'react';
import { connect } from 'react-redux';

import selectExpenses from '../selectors/expenses';
import ExpenseItem from './ExpenseItem';

export const ExpenseList = (props) => (
  <div className='container'>
    <div className='list-header'>
      <div className='show-for-mobile'>Expenses</div>
      <div className='show-for-desktop'>Expense</div>
      <div className='show-for-desktop'>Amount</div>
    </div>

    {props.expenses.length === 0 ? (
      <div className='list-body list-message'>
        <span>No expenses</span>
      </div>
    ) : (
      props.expenses.map((expense) => {
        return <ExpenseItem key={expense.id} {...expense} />;
      })
    )}
    <br />
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

export default connect(mapStateToProps)(ExpenseList);
