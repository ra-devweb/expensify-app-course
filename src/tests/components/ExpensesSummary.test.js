import React from 'react';
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/ExpensesSummary';

test('Should rendering ExpensesSummary correctly', () => {
  const wrapper = shallow(<ExpensesSummary />);

  expect(wrapper).toMatchSnapshot();
});

test('Should rendering ExpensesSummary correctly with single expense', () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={1} expensesTotal={255} />
  );

  expect(wrapper).toMatchSnapshot();
});

test('Should rendering ExpensesSummary correctly with multiple expenses', () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={3} expensesTotal={255366777} />
  );

  expect(wrapper).toMatchSnapshot();
});
