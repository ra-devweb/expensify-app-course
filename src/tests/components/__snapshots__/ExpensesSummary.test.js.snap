// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should rendering ExpensesSummary correctly 1`] = `
<div
  className="summary"
>
  <div
    className="container"
  >
    <h1
      className="summary__title"
    >
      Viewing 
      <span />
       
      expenses
       totalling 
      <span>
        0.00 $
      </span>
    </h1>
    <div
      className="summary__box-btn"
    >
      <Link
        className="summary__link btn"
        to="/create"
      >
        ✎ Add expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should rendering ExpensesSummary correctly with multiple expenses 1`] = `
<div
  className="summary"
>
  <div
    className="container"
  >
    <h1
      className="summary__title"
    >
      Viewing 
      <span>
        3
      </span>
       
      expenses
       totalling 
      <span>
        2,553,667.77 $
      </span>
    </h1>
    <div
      className="summary__box-btn"
    >
      <Link
        className="summary__link btn"
        to="/create"
      >
        ✎ Add expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`Should rendering ExpensesSummary correctly with single expense 1`] = `
<div
  className="summary"
>
  <div
    className="container"
  >
    <h1
      className="summary__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling 
      <span>
        2.55 $
      </span>
    </h1>
    <div
      className="summary__box-btn"
    >
      <Link
        className="summary__link btn"
        to="/create"
      >
        ✎ Add expense
      </Link>
    </div>
  </div>
</div>
`;
